### PR TITLE
argv: move formulae to cli/args

### PR DESCRIPTION
**Pull request by GauthamGoli**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----
#5730
Please do not merge until #6411 gets merged.

### CODE REVIEW

File: Library/Homebrew/cli/args.rb [hunk 1]
```diff
@@ -12,6 +12,135 @@ def initialize(argv:)
...
+          else
+            Formulary.find_with_priority(name, spec)
+          end
+        end.uniq(&:name)
+      end
+
+      def resolved_formulae
+        require "formula"
+        # TODO: use @instance variable to ||= cache when moving to CLI::Parser
+        (downcased_unique_named - casks).map do |name|
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cli/args.rb]: `@resolved_formulae ||= (downcased_unique_named - casks).map do |name|` and remove TODO

File: Library/Homebrew/cli/args.rb [hunk 2]
```diff
@@ -12,6 +12,135 @@ def initialize(argv:)
...
+          Formulary.resolve(name, spec: spec(nil))
+        end.uniq(&:name)
+      end
+
+      def casks
+        # TODO: use @instance variable to ||= cache when moving to CLI::Parser
+        downcased_unique_named.grep HOMEBREW_CASK_TAP_CASK_REGEX
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cli/args.rb]: `@casks ||= downcased_unique_named.grep HOMEBREW_CASK_TAP_CASK_REGEX` and remove TODO

File: Library/Homebrew/cli/args.rb [hunk 3]
```diff
@@ -12,6 +12,135 @@ def initialize(argv:)
...
+      end
+
+      def kegs
+        require "keg"
+        require "formula"
+        require "missing_formula"
+        # TODO: use @instance variable to ||= cache when moving to CLI::Parser
+        downcased_unique_named.map do |name|
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cli/args.rb]: `@kegs ||= downcased_unique_named.map do |name|` and remove TODO

File: Library/Homebrew/cli/parser.rb [hunk 4]
```diff
@@ -10,7 +10,7 @@
 module Homebrew
   module CLI
     class Parser
-      attr_reader :processed_options, :hide_from_man_page
+      attr_reader :processed_options, :hide_from_man_page, :args
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cli/parser.rb]: Shouldn't need to add this given `@args` usage below.`

File: Library/Homebrew/cli/args.rb (hunk 1 above)
Comment (GauthamGoli) [reply Library/Homebrew/cli/args.rb]: I've already tried exactly this but it was breaking a lot of tests. Let me try again.

Comment (MikeMcQuaid) [reply Library/Homebrew/cli/args.rb]: You may need to add a `Homebrew.args.reset` or similar which will reset the instance variables and have it in https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/spec_helper.rb#L135-L208

In general there's some ARGV stuff there that may need similar code for Homebrew.args

Comment (GauthamGoli) [reply Library/Homebrew/cli/args.rb]: I think we've hit a cyclic dependency here. 
[formula_options](https://github.com/Homebrew/brew/pull/6433/files#diff-e0e03670c451d934ae1d79adfff63cacL160) method is used to allow extra arguments for formulae see - [deps.rb#L55](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/cmd/deps.rb#L55) and [install.rb#L90](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/cmd/install.rb#L90) but formula options itself needs to know what is passed on command line to define what options can be allowed on command line!

Comment (MikeMcQuaid) [reply Library/Homebrew/cli/args.rb]: In that case you might want to allow e.g. `ARGV` to be passed in in some circumstances for initialisation so it's not operating on itself? Would that work?

Comment (GauthamGoli) [reply Library/Homebrew/cli/args.rb]: Yeah, if we limit the usage of `ARGV` to this one method we could make this work.

Comment (MikeMcQuaid) [reply Library/Homebrew/cli/args.rb]: I would suggest it's passed in as an argument (a default one if necessary) to make testing easier and just treated as an Array. The issue we're trying to address isn't so much the use of `ARGV` but removing our patches on `ARGV`.

**Comment by GauthamGoli** [comment]
@MikeMcQuaid  https://github.com/Homebrew/brew/runs/276504778#step:5:373
Its weird that `--ignore-dependencies` is being passed even without the changes in this PR, but its causing the test to fail here.

**Comment by MikeMcQuaid** [comment]
@GauthamGoli Are you stuck/blocked on this now? If so, can take a look and try to fix the tests. If it's not being passed but showing up that sounds like something being cached between tests that needs reset.

**Comment by GauthamGoli** [comment]
`--ignore-dependencies` is being passed, but on master branch, the warning doesn't show up in the test.

**Comment by MikeMcQuaid** [comment]
> `--ignore-dependencies` is being passed, but on master branch, the warning doesn't show up in the test. So its not something about cache.

@GauthamGoli Maybe this needs rebased on `master`? I've changed that code recently.

**Comment by GauthamGoli** [comment]
Test fail even after rebase.

File: Library/Homebrew/cli/args.rb [hunk 5]
```diff
@@ -50,6 +51,112 @@ def flags_only
...
+          if arg.include?("/") || arg.end_with?(".tar.gz") || File.exist?(arg)
+            arg
+          else
+            arg.downcase
+          end
+        end.uniq
+      end
+
+      def spec(default = :stable)
+        if HEAD?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cli/args.rb]: It looks like this is not being detected properly.

**Comment by MikeMcQuaid** [comment]
> `--ignore-dependencies` is being passed, but on master branch, the warning doesn't show up in the test. So its not something about cache.

Had a look: the issue isn't that `--ignore-dependencies` is being passed but that the output is not matching `Cloning into` or `testball1/HEAD` because it's ignoring/filtering out `--HEAD` for some reason.

**Comment by GauthamGoli** [comment]
Got it! Thanks. Its the chicken egg problem of `args` not being populated.

**Comment by GauthamGoli** [comment]
Tests pass now. 
@MikeMcQuaid  I think the code might need a bit of cleanup/simplification; can you please review, thanks!

File: Library/Homebrew/cli/parser.rb [hunk 6]
```diff
@@ -141,6 +143,7 @@ def parse(cmdline_args = ARGV)
         @args_parsed = true
         @args.processed_options = @processed_options
         Homebrew.args = @args
+        Homebrew.args.args_parsed = @args_parsed
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cli/parser.rb]: Think it would be good to handle both the above at once and perhaps even return `@args, @args_parsed, @parser` all from this method to be able to make these `attr_reader` instead of `attr_accessor`.

File: Library/Homebrew/cli/args.rb [hunk 7]
```diff
@@ -50,6 +51,111 @@ def flags_only
...
+          if arg.include?("/") || arg.end_with?(".tar.gz") || File.exist?(arg)
+            arg
+          else
+            arg.downcase
+          end
+        end.uniq
+      end
+
+      def spec(default = :stable)
+        if (args_parsed && HEAD?) || cmdline_args.include?("--HEAD")
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cli/args.rb]: Should/could this logic live inside `HEAD?`?

File: Library/Homebrew/cli/args.rb [hunk 8]
```diff
@@ -50,6 +51,111 @@ def flags_only
...
+          :head
+        elsif (args_parsed && devel?) || cmdline_args.include?("--devel")
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cli/args.rb]: Same here.

**Comment by zachauten** [comment]
@GauthamGoli @MikeMcQuaid So I guess I didn't notice that this PR moved the kegs method to args.rb when I started on #6622, should I just close that one and suggest changes here for the files that use ARGV.kegs?

**Comment by GauthamGoli** [comment]
@zachauten  #6622 need not be closed, after this PR is merged, you can rebase on master.

File: Library/Homebrew/cli/args.rb (hunk 7 above)
Comment (GauthamGoli) [reply Library/Homebrew/cli/args.rb]: `HEAD?` is one of the keys in OpenStruct, so I don't think so it possible. 
We could extract this logic out to another private method named `HEAD`

File: Library/Homebrew/cli/parser.rb (hunk 6 above)
Comment (GauthamGoli) [reply Library/Homebrew/cli/parser.rb]: I'm not sure I understand this completely. Could you please add in an example code snippet here.

File: Library/Homebrew/cli/args.rb (hunk 7 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/cli/args.rb]: Yeh, a private method for this sounds good 👍

File: Library/Homebrew/cli/parser.rb [hunk 9]
```diff
@@ -141,6 +143,7 @@ def parse(cmdline_args = ARGV)
         @args_parsed = true
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cli/parser.rb]:
```suggestion
         @args.args_parsed = @args_parsed = true
```

File: Library/Homebrew/cli/parser.rb (hunk 6 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/cli/parser.rb]:
```suggestion
```

File: Library/Homebrew/cli/args.rb [hunk 10]
```diff
@@ -5,13 +5,14 @@
 module Homebrew
   module CLI
     class Args < OpenStruct
-      attr_accessor :processed_options
+      attr_accessor :processed_options, :args_parsed
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cli/args.rb]:
```suggestion
      attr_reader :processed_options, :args_parsed
```

Comment (GauthamGoli) [reply Library/Homebrew/cli/args.rb]: https://github.com/Homebrew/brew/pull/6433/files#diff-e0e03670c451d934ae1d79adfff63cacR144
cannot be executed with only `attr_reader`

Comment (MikeMcQuaid) [reply Library/Homebrew/cli/args.rb]: Ok but I think it would be better to have a custom method that allows them to only be set once and then be frozen/not allow future changes to be made rather than using `attr_accessor`

Comment (GauthamGoli) [reply Library/Homebrew/cli/args.rb]: @MikeMcQuaid What do you think about this approach?

File: Library/Homebrew/cli/args.rb [hunk 11]
```diff
@@ -5,16 +5,23 @@
...
       undef tap
 
       def initialize(argv:)
         super
         @argv = argv
+        @args_parsed = false
         @processed_options = []
       end
 
+      def processed_options=(processed_options)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cli/args.rb]: I like this a lot more 👍. How about making it `freeze_processed_options!(processed_options)` to make it clearer it can only be called once?

File: Library/Homebrew/cli/args.rb [hunk 12]
```diff
@@ -5,16 +5,23 @@
...
+        @processed_options.concat(processed_options)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cli/args.rb]:
```suggestion
        @processed_options += processed_options
```

File: Library/Homebrew/cli/args.rb [hunk 13]
```diff
@@ -5,16 +5,23 @@
...
+        @args_parsed = true
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cli/args.rb]: I'd suggest putting this last just to avoid it being set if e.g. the `freeze` fails.

File: Library/Homebrew/cli/args.rb (hunk 11 above)
Comment (GauthamGoli) [reply Library/Homebrew/cli/args.rb]: Yep, updated.

File: Library/Homebrew/cli/args.rb (hunk 12 above)
Comment (GauthamGoli) [reply Library/Homebrew/cli/args.rb]: Addressed.

File: Library/Homebrew/cli/args.rb (hunk 13 above)
Comment (GauthamGoli) [reply Library/Homebrew/cli/args.rb]: Done.

**Comment by MikeMcQuaid** [comment]
Thanks again for more great work @GauthamGoli 🎉

**Comment by GauthamGoli** [comment]
We need to watch this closely for a few days for any bugs. I will be watching the issues 👍
